File: packages/core/src/resilience/circuit-breaker.ts
import { createLogger } from '../logging/logger.js';
import { CircuitBreakerError } from '../errors.js';

const log = createLogger('CircuitBreaker');

export type CircuitState = 'CLOSED' | 'OPEN' | 'HALF_OPEN';

export interface CircuitBreakerConfig {
  /** 서킷 이름 (로그용) */
  name: string;
  /** OPEN 전환까지 연속 실패 횟수 (기본 5) */
  failureThreshold?: number;
  /** OPEN → HALF_OPEN 대기 시간 ms (기본 30초) */
  resetTimeoutMs?: number;
  /** HALF_OPEN에서 허용할 시도 횟수 (기본 1) */
  halfOpenAttempts?: number;
}

/**
 * 서킷 브레이커 — 외부 서비스 장애 시 빠른 실패.
 * CLOSED → (연속 실패) → OPEN → (타임아웃) → HALF_OPEN → (성공) → CLOSED
 */
export class CircuitBreaker {
  private state: CircuitState = 'CLOSED';
  private failures = 0;
  private lastFailureTime = 0;
  private halfOpenSuccesses = 0;

  private readonly failureThreshold: number;
  private readonly resetTimeoutMs: number;
  private readonly halfOpenAttempts: number;
  private readonly name: string;

  constructor(config: CircuitBreakerConfig) {
    this.name = config.name;
    this.failureThreshold = config.failureThreshold ?? 5;
    this.resetTimeoutMs = config.resetTimeoutMs ?? 30_000;
    this.halfOpenAttempts = config.halfOpenAttempts ?? 1;
  }

  getState(): CircuitState {
    return this.state;
  }

  /**
   * 서킷 브레이커를 통해 함수를 실행한다.
   * OPEN 상태면 즉시 에러를 던진다.
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (this.state === 'OPEN') {
      // 타임아웃 경과 시 HALF_OPEN으로 전환
      if (Date.now() - this.lastFailureTime >= this.resetTimeoutMs) {
        this.state = 'HALF_OPEN';
        this.halfOpenSuccesses = 0;
        log.info({ circuit: this.name }, 'Circuit half-open, allowing probe request');
      } else {
        throw new CircuitBreakerError(this.name);
      }
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  private onSuccess(): void {
    if (this.state === 'HALF_OPEN') {
      this.halfOpenSuccesses++;
      if (this.halfOpenSuccesses >= this.halfOpenAttempts) {
        this.state = 'CLOSED';
        this.failures = 0;
        log.info({ circuit: this.name }, 'Circuit closed (recovered)');
      }
    } else {
      this.failures = 0;
    }
  }

  private onFailure(): void {
    this.failures++;

    if (this.state === 'HALF_OPEN') {
      this.lastFailureTime = Date.now();
      this.state = 'OPEN';
      log.warn({ circuit: this.name }, 'Circuit re-opened (half-open probe failed)');
    } else if (this.failures >= this.failureThreshold) {
      this.lastFailureTime = Date.now();
      this.state = 'OPEN';
      log.warn({ circuit: this.name, failures: this.failures }, 'Circuit opened');
    }
  }

  /** 수동 리셋 (테스트/관리용) */
  reset(): void {
    this.state = 'CLOSED';
    this.failures = 0;
    this.halfOpenSuccesses = 0;
  }
}

File: backend/src/core/resilience/circuit_breaker.py
from __future__ import annotations

import time
from enum import Enum
from typing import Awaitable, Callable, TypeVar, Union

from src.core.errors import CircuitBreakerError
from src.core.logging.logger import get_logger

log = get_logger("CircuitBreaker")

T = TypeVar("T")


class CircuitState(str, Enum):
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"


class CircuitBreaker:
    """
    외부 서비스 장애 시 빠른 실패 패턴.
    CLOSED → (연속 실패) → OPEN → (타임아웃) → HALF_OPEN → (성공) → CLOSED
    """

    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        reset_timeout_ms: int = 30_000,
        half_open_attempts: int = 1,
    ) -> None:
        self.name = name
        self._failure_threshold = failure_threshold
        self._reset_timeout_ms = reset_timeout_ms
        self._half_open_attempts = half_open_attempts

        self._state = CircuitState.CLOSED
        self._failures = 0
        self._last_failure_time: float = 0.0
        self._half_open_successes = 0

    @property
    def state(self) -> CircuitState:
        return self._state

    async def execute(self, fn: Callable[[], Union[Awaitable[T], T]]) -> T:
        if self._state == CircuitState.OPEN:
            elapsed_ms = (time.monotonic() - self._last_failure_time) * 1000
            if elapsed_ms >= self._reset_timeout_ms:
                self._state = CircuitState.HALF_OPEN
                self._half_open_successes = 0
                self._failures = 0
                log.info("Circuit half-open, allowing probe request", circuit=self.name)
            else:
                raise CircuitBreakerError(self.name)

        try:
            import asyncio
            result = fn()
            if asyncio.iscoroutine(result):
                result = await result
            self._on_success()
            return result
        except Exception:
            self._on_failure()
            raise

    def _on_success(self) -> None:
        if self._state == CircuitState.HALF_OPEN:
            self._half_open_successes += 1
            if self._half_open_successes >= self._half_open_attempts:
                self._state = CircuitState.CLOSED
                self._failures = 0
                log.info("Circuit closed (recovered)", circuit=self.name)
        else:
            self._failures = 0

    def _on_failure(self) -> None:
        self._failures += 1
        if self._state == CircuitState.HALF_OPEN:
            self._last_failure_time = time.monotonic()
            self._state = CircuitState.OPEN
            log.warn("Circuit re-opened (half-open probe failed)", circuit=self.name)
        elif self._failures >= self._failure_threshold:
            self._last_failure_time = time.monotonic()
            self._state = CircuitState.OPEN
            log.warn("Circuit opened", circuit=self.name, failures=self._failures)

    def reset(self) -> None:
        """수동 리셋 (테스트/관리용)."""
        self._state = CircuitState.CLOSED
        self._failures = 0
        self._half_open_successes = 0
